PYIC-5597: Repurpose `reset-identity` Lambda

sammyiyke opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes
[PYIC-5597](https://govukverify.atlassian.net/browse/PYIC-5597): Repurpose `reset-identity` Lambda

### What changed

- ~~Removes reset-identity lambda - `ResetIdentity`~~
- Creates new reset-session-identity lambda
- Updates Journey map -- removes reset-identity lambda journeys
- Updates Journey map -- adds reset-session-identity lambda journeys
- Removes reset-identity lambda and related resources from AWS deploy template
- Adds to logic for the new `ResetSessionIdentity` lambda
- Adds tests for new `ResetSessionIdentity` lambda logic
-  Replaces `RESET_IDENTITY` with new `RESET_SESSION_IDENTITY` for `PYI_ESCAPE_ABANDON_M2B`
-  Replaces `RESET_IDENTITY` with new `RESET_SESSION_IDENTITY ` for `CONFIRM_DELETE_DETAILS_PAGE`

#### There would be a follow up PR to clean up the `ResetIdentity` lambda ---> [see comment](https://github.com/govuk-one-login/ipv-core-back/pull/1847#discussion_r1579473720)

<!-- Describe the changes in detail - the "what"-->

### Why did it change

With the change to how identities are stored in ipv-core using Session VC Storage, we need to update the reset-identity lambda to fit. 

We no longer need to reset VCs at the start of a journey as these are stored/replaced at the end of the journeys and all other VCs are lost once the session ends.

Hence the `ResetIdentityHandler` is redundant and needs to be replaced.

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5597](https://govukverify.atlassian.net/browse/PYIC-5597)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [ ] No environment variables or secrets were added or changed

<!-- Delete if changes DO NOT include new environment variables or secrets -->
- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks


[PYIC-5597]: https://govukverify.atlassian.net/browse/PYIC-5597?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PYIC-5597]: https://govukverify.atlassian.net/browse/PYIC-5597?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sammyiyke committed:
PYIC-5597 - New Reset IPV Session Lambda Boilerplate

Sets up the boilerplate needed to get the new `ResetIpvSession` lambda up.

sammyiyke committed:
PYIC-5597 - Implement logic for ResetIpvSessionHandler lambda

Adds to logic for the new ResetIpvSession lambda

Gets the session ID and clears the session credentials

sammyiyke committed:
PYIC-5597 - Implement tests for ResetIpvSessionHandler lambda logic

Adds ResetIpvSession lambda test logic

sammyiyke committed:
PYIC-5597 - Adds RESET_IPV_SESSION to state journey event

Replaces `RESET_IDENTITY` with new `RESET_IPV_SESSION` for `PYI_ESCAPE_ABANDON_M2B`

We need to trigger `RESET_IPV_SESSION` when a user drops out of the M2B journey after collected at least one VC

sammyiyke committed:
PYIC-5597 - Cleanup ResetIdentityHandler

This removes the Reset Identity lambda and tests

Removes reset-identity lambda and related resources from AWS deploy template

Also fixes test that are dependent on the reprove-identity flag which is now unused

Removes local running for deleted lambda

sammyiyke committed:
PYIC-5597 - Cleanup

Replaces `ResetIdentityFunctionArn` with `ResetIpvSessionFunctionArn` in aws deploy template

sammyiyke committed:
PYIC-5597 - Cleanup

Replaces `ResetIdentityFunctionArn` with `ResetIpvSessionFunctionArn` in aws deploy template

sammyiyke committed:
PYIC-5597 - Updates Journey Map

Updates Journey map to remove reset-identity lambda journeys

Add reset-ipv-session lambda journeys to map

sammyiyke committed:
PYIC-5597 - Refactor - Changed to ResetIpvSessionHandler.java to ResetSessionIdentity

Changed naming for new lambda and related tests and methods

sammyiyke committed:
PYIC-5597 - Removed reset-identity and ResetIdentityLambda

Added new ResetSessionIdentityLambda

sammyiyke committed:
PYIC-5597 - Removed unused exception

sammyiyke committed:
PYIC-5597 - Removed reset-gpg45-identity

Removed missed reset-gpg45-identity and related tests

sammyiyke committed:
PYIC-5597 - Fixes

AWS deploy failure fix

sammyiyke committed:
PYIC-5597 - SpotlessApply 😓

Joe-Edwards-GDS commented on deploy/journeyEngineStepFunction.asl.json:
```diff
@@ -25,11 +25,6 @@
-          "Variable": "$.journey",
-          "StringMatches": "/journey/reset-identity",
-          "Next": "ResetIdentityLambda"
-        },
```
Have you tried this in your dev environment? I think you'll need to add corresponding blocks for the new lambda as well

sammyiyke replied:
I did. Seems to work fine. I might be wrong but I think it is related to lambdas with "stops" on the journey.

Probably need to improve my understanding of the Journey system.

sammyiyke replied:
But it should be coming back with the new changes.

Joe-Edwards-GDS commented on deploy/template.yaml:
```diff
@@ -2613,6 +2523,74 @@ Resources:
+        - SSMParameterReadPolicy:
+            ParameterName: !Sub ${Environment}/core/*
+        - SQSSendMessagePolicy:
+            QueueName: !ImportValue AuditEventQueueName
```
I think you'll probably need CRUD permissions for some of the dynamodb tables here as well

Joe-Edwards-GDS commented on settings.gradle:
```diff
@@ -16,8 +16,8 @@ include "lambdas",
 		"lambdas:replay-cimit-vcs",
-		"lambdas:reset-identity",
 		"lambdas:store-identity",
+		"lambdas:reset-session-identity",
```
Do you think we can try and keep these alphabetical?

Joe-Edwards-GDS commented on local-running/src/main/java/uk/gov/di/ipv/coreback/handlers/JourneyEngineHandler.java:
```diff
@@ -93,8 +90,6 @@ private Map<String, Object> processJourneyStep(
             case "/journey/check-existing-identity" -> checkExistingIdentityHandler.handleRequest(
                     buildJourneyRequest(request, journeyStep), EMPTY_CONTEXT);
-            case "/journey/reset-identity" -> resetIdentityHandler.handleRequest(
-                    buildProcessRequest(request, processJourneyEventOutput), EMPTY_CONTEXT);
```
Similarly, this will need a corresponding addition for the new lambda.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -229,15 +219,6 @@ private JourneyResponse getJourneyResponse(
-                return buildForceResetResponse(ciScoringCheckResponse.orElse(null));
-            } else if (configService.enabled(RESET_IDENTITY)) {
-                return buildForceGpg45ResetResponse(ciScoringCheckResponse.orElse(null));
-            }
```
We can't quite just get rid of this I'm afraid - if the reprove identity flag is set, we still need to send a response here - we just don't need to reset the identity first.

I think the logic should be:
- If the reprove identity / reset identity flag is 'true', return `/journey/reprove-identity`
- In the journey map, route that to the `REPROVE_IDENTITY` entry point of `NEW_P2_IDENTITY`

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -441,13 +387,6 @@ private JourneyResponse buildNoMatchResponse(
-        if (!VcHelper.filterVCBasedOnProfileType(verifiableCredentials, GPG45).isEmpty()) {
-            LOGGER.info(
-                    LogHelper.buildLogMessage("Failed to match profile so resetting identity."));
-            sendAuditEvent(AuditEventTypes.IPV_IDENTITY_REUSE_RESET, auditEventUser);
```
If we're losing this audit event, we should also check with the D&A and TICF teams, to make sure they aren't using it for anything.

Joe-Edwards-GDS replied:
We've done this, and agreed this is fine :)

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -728,18 +727,6 @@ PYI_KBV_DROPOUT_M2B:
-    lambdaInput:
-      isUserInitiated: false
-      deleteOnlyGPG45VCs: true
-    mitigationStart: enhanced-verification
```
We'll need to move these `mitigationStart` events onto the next step in the journey now - otherwise we'll lose the corresponding audit event.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java:
```diff
@@ -88,7 +88,8 @@ public enum ErrorResponse {
     INVALID_JOURNEY_EVENT(1073, "Invalid journey event in input"),
-    FAILED_TO_STORE_IDENTITY(1074, "Failed to store identity");
+    FAILED_TO_STORE_IDENTITY(1074, "Failed to store identity"),
+    FAILED_TO_RESET_IPV_SESSION(1075, "Failed to reset ipv session");
```
Looks like this is only used in tests - can we delete?

Joe-Edwards-GDS reviewed (commented):
Looks like the right idea, but I'd definitely suggest running this in your dev environment, and ideally trying to run the integration tests against it as well.

sammyiyke committed:
PYIC-5597 - Removes single use ErrorResponse

Removes `FAILED_TO_RESET_IPV_SESSION` which is only used in a test case

sammyiyke committed:
PYIC-5597 - PR related changes

- Fix to make the gradle include list alphabetical

- Add `mitigationStart` to the next journey for M1 mitigation. Otherwise, we'll lose the corresponding audit event.

sammyiyke committed:
PYIC-5597 - PR related changes

- Adds CRUD permission for related dynamoDB tables

sammyiyke committed:
PYIC-5597 - PR review related changes

- Adds new logic for reprove identity flag

sammyiyke committed:
PYIC-5597 - PR review related changes

- Adds `ResetSessionIdentityLambda` to Journey engine

sammyiyke committed:
PYIC-5597 - PR review related changes

- Cleanup

sammyiyke committed:
PYIC-5597 - PR review related changes

- Cleanup on tests

sammyiyke committed:
PYIC-5597 - PR review related changes

- Renames `JOURNEY_RESET`
- Changes URL `/journey/reset-identity` to `/journey/reset-session-identity` to match new lambda

sammyiyke committed:
PYIC-5597 - PR review related changes

- Renames `JOURNEY_RESET`
- Changes URL `/journey/reset-identity` to `/journey/reset-session-identity` to match new lambda

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -228,14 +222,14 @@ private JourneyResponse getJourneyResponse(
+                LOGGER.info(
+                        LogHelper.buildLogMessage(
+                                "resetIdentity flag is enabled, reset users identity."));
+                return JOURNEY_RESET_SESSION_IDENTITY;
```
This doesn't actually need to reset the session identity, it can link straight through to `NEW_P2_IDENTITY` / `REPROVE_IDENTITY`.

I'd probably call the event `/journey/reprove-identity` or similar

Joe-Edwards-GDS replied:
I think this comment still applies :)

Joe-Edwards-GDS reviewed (commented):
This looks good, just one small comment, and we need to see what comes out of the audit event conversations :)

sammyiyke committed:
PYIC-5597 - Rebase and Cleanup

Joe-Edwards-GDS commented on deploy/template.yaml:
```diff
@@ -1673,93 +1670,6 @@ Resources:
       LogGroupName: !Ref CheckExistingIdentityFunctionLogGroup
 
-
-  ResetIdentityFunction:
```
Do you think there's any risk of this breaking active sessions? I'm wondering if it might be safer to split this into two PRs:
- First add the new lambda and make all the journey map changes
- Second delete the old lambda

cc @Wynndow

Wynndow replied:
Yeah, I think it'd be safe to drain the lambda before removing it.

sammyiyke replied:
Brought back the `resetIdentity` lambda as part of this PR. Would create a second PR for the clean up after this is merged

Joe-Edwards-GDS replied:
Thanks!

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/initial-journey-selection.yaml:
```diff
@@ -61,35 +55,7 @@ CHECK_EXISTING_IDENTITY:
-    lambda: reset-identity
-    lambdaInput:
-      isUserInitiated: false
-      deleteOnlyGPG45VCs: false
```
Can you remove this state entirely? I'm not sure it's valid at the moment

sammyiyke committed:
PYIC-5597 - PR Related Fixes

Removes `JOURNEY_RESET_SESSION_IDENTITY` as there is no need to reset the session identity.

Links straight to `NEW_P2_IDENTITY`

sammyiyke committed:
PYIC-5597 - PR Related Fixes

Removes unused state `RESET_IDENTITY`

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -229,7 +226,7 @@ private JourneyResponse getJourneyResponse(
                         LogHelper.buildLogMessage(
                                 "resetIdentity flag is enabled, reset users identity."));
-                return JOURNEY_RESET_SESSION_IDENTITY;
+                return JOURNEY_IPV_GPG45_MEDIUM;
```
Sorry @sammyiyke  - this needs to go to a slightly different place. We still need to display the extra reprove identity screen, so it needs to go to `NEW_P2_IDENTITY`/`REPROVE_IDENTITY` (rather than `NEW_P2_IDENTITY`/`START`)

sammyiyke committed:
PYIC-5597 - Updates secrets

sammyiyke committed:
PYIC-5597 - PR related Fixes

Fixed reprove identity journey to point to `NEW_P2_IDENTITY/REPROVE_IDENTITY` rather than `NEW_P2_IDENTITY/START`

sammyiyke committed:
PYIC-5597 - Adds `ResetSessionIdentityFunctionArn` to `JourneyEngineStepFunction`

sammyiyke committed:
PYIC-5597 - Temp.. Returns the `ResetIdentity` lambda.

"it'd be safe to drain the lambda before removing it."
